common/org.testeditor.dsl.common.ui/src/org/testeditor/dsl/common/ui/refactoring/NamedElementRenameParticipant.xtend

```diff
@@ -0,0 +1,139 @@
+
+@Accessors(PROTECTED_GETTER) // for testing
+class NamedElementRenameParticipant extends RenameParticipant {
+
+	@Inject PartHelper partHelper
+	@Inject IResourceSetProvider resourceSetProvider
+	@Inject XtextDocumentProvider documentProvider
+
+	val Set<String> fileExtensions
+	@Accessors(PUBLIC_GETTER) val String name
+	IFile element
+
+	@Inject
+	new(@Named(FILE_EXTENSIONS) String fileExtensions, @Named(LANGUAGE_NAME) String languageName) {
+		this.fileExtensions = fileExtensions.split(",").map[toLowerCase].toSet
+		this.name = languageName.split("\\.").last + " Renaming"
+	}
+
+	override checkConditions(IProgressMonitor pm, CheckConditionsContext context) throws OperationCanceledException {
+		// TODO check the validity of the new name
+	}
+
+	/** 
+	 * Creates the change for the name refactoring.
+	 * Two cases need to be distinguished:
+	 * <ul>
+	 * 	<li>Element is opened in an Xtext editor</li>
+	 * 	<li>Element is not opened and refactored using the navigator</li>
+	 * </ul>
+	 */
+	override createPreChange(IProgressMonitor pm) throws CoreException, OperationCanceledException {
+		val editor = partHelper.findEditor(element)
+		if (editor instanceof XtextEditor) {
+			return createEditorChange(editor)
+		} else {
+			return createTextFileChange
+		}
+	}
+
+	override createChange(IProgressMonitor pm) throws CoreException, OperationCanceledException {
+		// do nothing here, we want to create "pre changes" as we want the name refactoring
+		// to happen before the resource is moved
+		return null
+	}
+
+	/**
+	 * Creates a name replacement for a resource with an open editor.
+	 */
+	private def Change createEditorChange(XtextEditor editor) {
+		val namedElement = getNamedElement(editor.document)
+		val nameReplacement = createNameReplacement(namedElement)
+		val shouldSave = !editor.isDirty
+		return new EditorDocumentChange("Change name", editor, shouldSave) => [
+			edit = nameReplacement
+			textType = element.fileExtension
+		]
+	}
+
+	/**
+	 * Creates a name replacement for a resource with no open editor.
+	 */
+	private def Change createTextFileChange() {
+		// retrieve the document
+		val input = new FileEditorInput(element)
+		documentProvider.connect(input)
+		try {
+			val document = documentProvider.getDocument(input) as IXtextDocument
+			val namedElement = getNamedElement(document)
+			val nameReplacement = createNameReplacement(namedElement)
+
+			// create the change
+			return new TextFileChange("Change name", element) => [
+				saveMode = TextFileChange.FORCE_SAVE
+				edit = nameReplacement
+				textType = element.fileExtension
+			]
+		} finally {
+			documentProvider.disconnect(input)
+		}
+	}
+
+	@VisibleForTesting
+	protected def ReplaceEdit createNameReplacement(NamedElement namedElement) {
+		val node = namedElement.findNodesForFeature(NAMED_ELEMENT__NAME).head
+		val newName = arguments.newName.replaceFirst('''\.(«fileExtensions.join("|")»)''', "")
```

Comment:
What if I rename `MyTest.tcl` to `MyAwesomeTest.TCL`. The replacement performed here will not cover this case.